illumos-utils/src/svcs.rs

```diff
@@ -58,16 +57,95 @@ impl Svcs {
     pub fn new() -> Self {
-        Self { services: vec![], time_of_status: None }
+        Self { services: vec![], errors: vec![], time_of_status: None }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.services.is_empty() && self.time_of_status == None
+        self.services.is_empty()
+            && self.errors.is_empty()
+            && self.time_of_status == None
+    }
+
+    // This method is only used when the target OS is "illumos". It is not
+    // marked as a configuration option based on target OS because it is not
+    // Illumos specific itself. Additionally, we would not be able to run the
+    // unit tests on any other OS if we mark it to be used on Illumos only.
+    // We mark it as unused instead.
+    #[allow(dead_code)]
+    fn parse(log: &Logger, data: &[u8]) -> Self {
+        let mut services = vec![];
+        let mut errors = vec![];
+        if data.is_empty() {
+            return Self { services, errors, time_of_status: Some(Utc::now()) };
+        }
+
+        // Example of the reponse from running `svcs -Za -H -o state,fmri,zone`
+        //
+        // legacy_run     lrc:/etc/rc2_d/S20sysetup                          global
+        // maintenance    svc:/site/fake-service:default                     global
+        // disabled       svc:/network/tcpkey:default                        global
+        // disabled       svc:/system/omicron/baseline:default               global
+        // online         svc:/milestone/sysconfig:default                   global
+        let s = String::from_utf8_lossy(data);
+        let lines = s.trim().lines();
+        for line in lines {
+            let line = line.trim();
+            let mut svc = line.split_whitespace();
+
+            if let Some(state) = svc.next() {
+                // Only attempt to parse a service that is in maintenance.
+                match SvcState::from(state.to_string()) {
+                    SvcState::Maintenance => {
+                        // This is a new service, wipe the slate clean
+                        let mut current_svc = SvcInMaintenance::new();
+                        if let Some(fmri) = svc.next() {
+                            current_svc.fmri = fmri.to_string()
+                        }
+
+                        if let Some(zone) = svc.next() {
+                            current_svc.zone = zone.to_string()
+                        } else {
+                            // We only need to collect an error here. If the
+                            // previous svc.next() was `None`, this one will be
+                            // `None` as well.
+                            errors.push(format!(
+                                "Unexpected output line: {line}"
+                            ));
+                            error!(
+                                log,
+                                "unable to parse; output line missing zone: {line}",
+                            );
+                        }
```

Comment:
nitpicky: it looks like if there is no FMRI _or_ zone, we fall through and log "output line missing zone" even if the FMRI is also missing, which may be not quite the most correct description of the error?